optuna/importance/_mean_decrease_impurity.py

```diff
@@ -82,27 +85,27 @@ def evaluate(
         trans = _SearchSpaceTransform(distributions, transform_log=False, transform_step=False)
 
         n_trials = len(trials)
-        trans_params = numpy.empty((n_trials, trans.bounds.shape[0]), dtype=numpy.float64)
-        trans_values = numpy.empty(n_trials, dtype=numpy.float64)
+        self._trans_params = numpy.empty((n_trials, trans.bounds.shape[0]), dtype=numpy.float64)
+        self._trans_values = numpy.empty(n_trials, dtype=numpy.float64)
 
         for trial_idx, trial in enumerate(trials):
-            trans_params[trial_idx] = trans.transform(trial.params)
-            trans_values[trial_idx] = trial.value if target is None else target(trial)
+            self._trans_params[trial_idx] = trans.transform(trial.params)
+            self._trans_values[trial_idx] = trial.value if target is None else target(trial)
 
         encoded_column_to_column = trans.encoded_column_to_column
 
-        if trans_params.size == 0:  # `params` were given but as an empty list.
+        if self._trans_params.size == 0:  # `params` were given but as an empty list.
             return OrderedDict()
 
         forest = self._forest
-        forest.fit(trans_params, trans_values)
+        forest.fit(self._trans_params, self._trans_values)
         feature_importances = forest.feature_importances_
         feature_importances_reduced = numpy.zeros(len(distributions))
         numpy.add.at(feature_importances_reduced, encoded_column_to_column, feature_importances)
 
         param_importances = OrderedDict()
-        param_names = list(distributions.keys())
+        self._param_names = list(distributions.keys())
         for i in feature_importances_reduced.argsort()[::-1]:
-            param_importances[param_names[i]] = feature_importances_reduced[i].item()
+            param_importances[self._param_names[i]] = feature_importances_reduced[i].item()
 
         return param_importances
```

Comment:
Thank you for your PR! 
As also mentioned by @nzw0301, we don't actually need to introduce these member fields. Instead of having `_trans_params`, `_trans_values` and `_param_names` as member fields, we could make another function `_evaluate` returning `(param_importances, trans_params, trans_values, param_names)` in `MeanDecreaseImpurityImportanceEvaluator`  and call that function from both `ShapleyImportanceEvaluator.evaluate()` and `MeanDecreaseImpurityImportanceEvaluator.evaluate()`.

Reply:
I would like to suggest spliting the current evaluation methods into i) pre-processing part that returns `trans, trans_params, trans_values` and ii) the computing importance score part because the evaluator classes including this shap class call this parameter transformation. The first part might be the static method and the second part calls the first pre-processing method. 
I'm still not sure this approach is optimal.

Reply:
hi @contramundum53 @nzw0301 I do understand your concerns about separating the concerns and removing the class members. The problem is the implementation of `MeanDecreaseImpurityImportanceEvaluator.evaluate()` is so intertwined with many states and variables. 

If we were to have a separate method to retrieve the parameters from the trials, then returning the tuple of `(trans, trans_params, trans_values)` isn't sufficient, we need also the `distribution` which is used at the end of the function. In addition, there are some exceptions raising and early returns in-between that need to be properly re-transformed. I'm not convinced that there is much to gain by doing so. 

Will the code be more modularized? probably. But one might argue that having the class encapsulate the state information is itself a good way to modularize the code, which is the current situation now.

Reply:
Right, but then the pre-processing takes `distributions` as an argument to be reused after random forest fitting. This does not need to return an empty `OrderedDict`; we might return it before and after preprocessing.

What do you think @g-votte? I would like to know whether the first approach suggested in https://github.com/optuna/optuna/pull/3507#issuecomment-1110439225 is the current implementation or not. Specifically, can we introduce new private attributes to `MeanDecreaseImpurityImportanceEvaluator` for `ShapleyImportanceEvaluator`.

Reply:
hi @nzw0301 @g-votte @contramundum53 I've given it another thought. 

First of all, I think it is a good thing to keep the class `MeanDecreaseImpurityImportanceEvaluator` **stateful**, rather than insisting on having a _stateless_ `evaluate()` function. One would like to look into the evaluator after the evaluation if there is anything wrong, or simply the state information could be useful for other purposes (e.g. pass it on to `ShapleyImportanceEvaluator`).

If you agree with the above point, then we can address the points that make you guys uneasy (including myself), namely 1). we access directly the attributes of `MeanDecreaseImpurityImportanceEvaluator` (although Python doesn't have any access control anyway). 2). it is not entirely necessary to evaluate `MeanDecreaseImpurityImportanceEvaluator` in `ShapleyImportanceEvaluator`. 

To address the above two concerns, here is what I propose (hopefully the code is self-explanatory): 

```
       # Retrieve the trial parameters and fit a RandomForest model out of them.
       model = self._backend_evaluator.fit(study=study, params=params, target=target)
       
       # Retrieve the trial parameters.
       (trans_params, trans_values, param_names)  = self._backend_evaluator.get_params()
```

PS: the above APIs are inspired by the `sklearn.BaseEstimator`.

Reply:
hi @nzw0301 @g-votte @contramundum53 any feedback on the above proposal?

Reply:
Sorry, we were on a vacation.
Thanks, I agree with your proposal.
One minor comment: I think it would be more readable and robust to make the return value of `get_params()` a `dict` instead of a `tuple`.

Reply:
Sorry for coming late. I think the current PR reflects what I suggested in https://github.com/optuna/optuna/pull/3507#issuecomment-1110439225 .

As mentioned by @nzw0301 in https://github.com/optuna/optuna/pull/3507#pullrequestreview-970496123, we'd need some followup refactoring to extract the common logic among importance evaluators and to get rid of unnecessary calculation with the current approach. #3552 will shortly address those problems.